### ISSUE TITLE
Limiting mob breeding / mobs per chunk, etc

### ISSUE THREAD

**Issue by suresttexas00** [benbaptist/minecraft-wrapper#149]
I have had some players overbreed their chickens and cows.  Invariably, i'll notice the "can't keep up messages" where the server is skipping more than a hundred ticks (with only one player or so online)... and I will find a chicken farm with 400 chickens and 100 cows.... Is there a way to count an entity type inside a world,chunk, or radius... figure out if some count is exceeded and do a kill @e[(xyz coords of player),r=100ish,type=Chicken,c=?]... It's that "c" number i need help setting... i can't just kill all the entities.. just want to kill the ones above a certain count.  I have this problem at zombie spawners too.

Basically, this all hinges on "counting" a type of entity in a given area (around each player?).

**Comment by suresttexas00** [comment]
I saw that the /testfor command can count entities.  not sure I want to spam the console with 500 "Found Chicken"'s, but if wrapper could read a testfor result, that might do the trick.

**Comment by suresttexas00** [comment]
looks like the proxy.py and entities.py code is doing some limited tracking of entities... we might be able to get some kind of object that the plugins can access to do entity work.

**Comment by JasonBristol** [comment]
+1 for plugin support via api

**Comment by suresttexas00** [comment]
Progress here.

The world entities object is now tracking entities by spawn, movement, and destruction.

I just need to add a method to kill a specific entity by their EID.  Then we can implement api things to probe their concentrations and thin excess eitities.

**Comment by suresttexas00** [comment]
This will be easier than I thought because entities (properly speaking) should not really be a world item (even though that is how we have implemented them).  They should be a player item.  However, I have a world-wide construct in place to manage these by having assigned each entity to its respective player.

instead of searching chunk concentrations, we can simply use the player(s) as the fulcrum for entity management code.

**Comment by suresttexas00** [comment]
I have the following functions now in api.minecraft.world:
```
def getEntityByEID(self, eid)  # part of original wrapper
 ...
def countActiveEntities(self):
    """ return a count of all entities. """
def getEntityInfo(self, eid):
    """ get dictionary of info on the specified EID.  Returns None if fails """
def countEntitiesInPlayer(self, playername):
    """returns a list of entity info dictionaries """
def existsEntityByEID(self, eid):
    """ A way to test whether the specified eid is valid """
def killEntityByEID(self, eid, dropitems=False, finishstateof_domobloot=True, count=1):
    """ takes the entity by eid and kills the first entity of that type around those coordinates """
```

I don't think the API really needs anything else.  Plugins can do the actual management work.  Anything else that should be in the API ?

**Comment by suresttexas00** [comment]
I think a new config would be neat.  Have a section called 'entity-limits'.  within that section, a player can put things like `"Cow": 100`.   Use the standard minecraft entity names to allow any entity to be configured.

**Comment by suresttexas00** [comment]
Added section in Config:

```
    "Gameplay": {
        "use-timer-tick-event": False,  # not recommended.  1/20th of a second timer option for plugin use. May
                                        # impact wrapper performance negatively.
        "enable-entity-controls": False,  # enable entity controls.
        "thinning-frequency": 10,  # how often thinning of mobs runs, in seconds
        "thinning-activation-threshhold": 100,  # when TOTAL mobs are below this number, thinning is skipped entirely
        "thin-any-mob": 50,  # any mob count above this number gets thinned.
        "thin-Cow": 30,  # Example, keeps Cows < 30.  Name must match exactly.  Overrides 'thin-any-mob'.
        "thin-Sheep": 30,
        "thin-Chicken": 30
    },
```
